File: prettier-plugin-csharpier/src/Printer/Types/QueryBody.ts
import { Doc } from "prettier";
import { PrintMethod } from "../PrintMethod";
import { SyntaxTreeNode } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";
import { printQueryContinuation, QueryContinuationNode } from "./QueryContinuation";

export interface QueryBodyNode extends SyntaxTreeNode<"QueryBody"> {
    clauses: SyntaxTreeNode[];
    selectOrGroup: SyntaxTreeNode;
    continuation?: QueryContinuationNode;
}

export const printQueryBody: PrintMethod<QueryBodyNode> = (path, options, print) => {
    const node = path.getValue();
    const parts: Doc[] = [];
    parts.push(join(line, path.map(print, "clauses")));
    if (node.clauses.length > 0) {
        parts.push(line);
    }
    parts.push(path.call(print, "selectOrGroup"));
    if (node.continuation) {
        // TODO indent when there is a group by before the into?
        parts.push(" ", path.call(o => printQueryContinuation(o, options, print), "continuation"));
    }

    return concat(parts);
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/QueryBody.cs
namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class QueryBody
{
    public static Doc Print(QueryBodySyntax node, FormattingContext context)
    {
        Doc clausesSeparator = (context.NewLineBetweenQueryExpressionClauses == true) && (node.Clauses.Count > 0)
            ? Doc.HardLine
            : (context.NewLineBetweenQueryExpressionClauses == false ? " " : Doc.Line);
        var docs = new List<Doc>
        {
            clausesSeparator,
            Doc.Join(clausesSeparator, node.Clauses.Select(o => Node.Print(o, context)))
        };
        if (node.Clauses.Count > 0)
        {
            docs.Add(Doc.Line);
        }

        docs.Add(Node.Print(node.SelectOrGroup, context));
        if (node.Continuation != null)
        {
            docs.Add(" ", QueryContinuation.Print(node.Continuation, context));
        }

        return Doc.Concat(docs);
    }
}
